### DISCUSSION TITLE
Guided Capstone Project - MERN - Meeting App - Meet in Middle

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/mern#205] [Tasks]
Let's start with the initial steps and we can proceed with more as per your requirement. I'm starting with the first two points - "OTP based authentication" and "Creating new meetings and sending invites to people via SMS".

### Prerequisite:
- Basic understanding of JavaScript, Node.js, React, MongoDB, and Express.
- Node.js, npm, MongoDB installed.
- An AWS account.
- An account with an SMS gateway like Twilio.
- Setup a new React project and a Node.js server.

#### Frontend (React)
a. **Landing Screen**: Create UI with options for sign-in and sign-up. 
b. **Sign-up Screen**: Create a form to take user name, mobile number, and profile picture.
c. **OTP Screen**: Create a form to accept OTP input.

##### Code Snippet Example (OTP Screen)
```jsx
import React, { useState } from "react";
// ... other imports ...

function OtpScreen() {
    const [otp, setOtp] = useState("");
    
    const handleOtpChange = (e) => {
        let value = e.target.value;
        
        if (/[^0-9]/.test(value)) {
            // Alert if entered non-digit
            alert("Only digits allowed!");
        } else {
            setOtp(value.slice(0, 4));  // Keep OTP to max 4 digits
        }
    };
    
    //... other methods...

    return (
        <div>
            <input type="text" value={otp} onChange={handleOtpChange} maxLength="4" />
            {/* ... other elements ... */}
        </div>
    );
}

export default OtpScreen;
```

#### Backend (Node.js & Express)
a. **Generating OTP**: Generate a 4-digit OTP on the backend server.
b. **Sending OTP**: Send the OTP via SMS using Twilio or any other provider.
c. **Verifying OTP**: Validate the OTP received from the frontend.

##### Code Snippet Example (Sending OTP)
```js
const express = require("express");
const Twilio = require("twilio");

const app = express();
const twilioClient = Twilio('ACCOUNT_SID', 'AUTH_TOKEN');

app.post("/send-otp", (req, res) => {
    const otp = Math.floor(1000 + Math.random() * 9000);  // Generate OTP
    
    twilioClient.messages.create({
        body: `Your OTP is: ${otp}`,
        to: `+91${req.body.phoneNumber}`, // Replace with user's phone number
        from: '+1234567890' // Replace with your Twilio number
    })
    .then((message) => console.log(message.sid))
    .catch((error) => console.error(error));
    
    //... other logics like storing otp for validation ...
});

//... other routes...

app.listen(3000, () => {
    console.log("Server is running on port 3000");
});
```

#### Frontend
a. **Creating Meeting**: Implement a form to gather meeting data (name, time, participants, etc.) and POST it to the backend.
b. **Sending Invites**: Implement functionality to send SMS invitations after meeting creation.

#### Backend
a. **Storing Meeting Data**: Save the meeting data to MongoDB including the details of invitees.
b. **Sending SMS**: Upon creation of a meeting, send SMS to invitees with a link to join the meeting.

##### Code Snippet Example (Storing Meeting Data & Sending Invites)
```js
// Assuming body-parser and mongoose (connected to MongoDB) are configured...

app.post("/create-meeting", (req, res) => {
    // Extract meeting data from req.body
    const { meetingName, time, participants } = req.body;
    
    // Save meeting to DB
    // Code for saving to MongoDB...

    // Send SMS invites
    participants.forEach(participant => {
        twilioClient.messages.create({
            body: `You are invited to ${meetingName} at ${time}. Join here: [LINK]`,
            to: `+91${participant.phoneNumber}`,
            from: '+1234567890'
        })
        .then((message) => console.log(message.sid))
        .catch((error) => console.error(error));
    });

    //... other logic...
});

//... other routes...
```

### Further Steps:
- We need to progress towards implementing the other steps/features of the app like joining a meeting, finding mid-meeting places, and building home and profile screens.
- Each step would also include ensuring secure API calls, maintaining session states, and real-time updates using WebSocket for notifying users in real-time when a new user joins the meeting, or a meeting detail changes.
- Further development should also consider robust error handling and user-friendly UI/UX designs.

---

#### Frontend
a. **Meeting Link**: Create a dynamic route to handle meeting links, e.g., `/meeting/:meetingId`.
b. **Meeting Screen**: Display meeting details and participant information.

#### Backend
a. **Retrieve Meeting Data**: Upon a request to join, fetch meeting data from MongoDB and send it to the frontend.

##### Code Snippet Example (Meeting Screen - React)
```jsx
import React, { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';

function MeetingScreen() {
    const { meetingId } = useParams();
    const [meetingData, setMeetingData] = useState(null);

    useEffect(() => {
        fetch(`http://localhost:3000/meeting/${meetingId}`)
            .then((response) => response.json())
            .then((data) => setMeetingData(data))
            .catch((error) => console.error('Error:', error));
    }, [meetingId]);

    return (
        <div>
            {meetingData ? (
                <>
                    <h1>{meetingData.name}</h1>
                    <p>Time: {new Date(meetingData.time).toLocaleString()}</p>
                    {/* ... Display other meeting data ... */}
                </>
            ) : (
                <p>Loading...</p>
            )}
        </div>
    );
}

export default MeetingScreen;
```

##### Code Snippet Example (Fetching Meeting Data - Node.js/Express)
```js
//... other imports and setups...

app.get("/meeting/:meetingId", (req, res) => {
    const { meetingId } = req.params;
    
    // Fetch meeting data from MongoDB using meetingId
    // Assume Meeting is your MongoDB model
    Meeting.findById(meetingId)
        .then((meeting) => {
            if (!meeting) {
                res.status(404).json({ error: 'Meeting not found' });
            } else {
                res.json(meeting);
            }
        })
        .catch((error) => res.status(500).json({ error: 'Server error' }));
});

//... other routes...
```

#### Frontend
a. **Location Sharing**: Implement functionality for participants to share their current locations.
b. **Map Display**: Use a map service like Google Maps API to display user locations and suggest mid-meeting places.

#### Backend
a. **Storing Locations**: Store the shared locations in the database.
b. **Finding Mid-point**: Implement an algorithm to find a central point among all participants.
c. **Finding Venues**: Utilize APIs like Google Places API to find venues near the central point.

##### Code Snippet Example (Finding Mid-point - Node.js/Express)
```js
//... other imports and setups...

// Assume receivedLocations is an array of {latitude, longitude} objects
// from all meeting participants.

const findMidPoint = (receivedLocations) => {
    let avgLat = 0, avgLng = 0;
    receivedLocations.forEach(location => {
        avgLat += location.latitude;
        avgLng += location.longitude;
    });
    avgLat /= receivedLocations.length;
    avgLng /= receivedLocations.length;
    
    return { latitude: avgLat, longitude: avgLng };
};

app.post("/find-midpoint", (req, res) => {
    const receivedLocations = req.body.locations;
    
    if (!receivedLocations || receivedLocations.length === 0) {
        res.status(400).json({ error: 'No locations provided' });
        return;
    }

    const midPoint = findMidPoint(receivedLocations);
    
    // Further you can find nearby places using Google Places API and send back to frontend...
    
    res.json(midPoint);
});

//... other routes...
```

### Further Steps:
- You'd need to implement logic for fetching nearby places based on the midpoint calculated.
- Implementing chat functionality during the meeting using WebSockets for real-time bi-directional communication between participants.
- Handling user data management, like updating profile information, managing friend lists, etc.

Each step contains lots of details, so feel free to ask for further elaboration on any specific part. You can also implement these steps one by one, testing thoroughly before moving to the next to ensure a stable development process.

---

#### Frontend
a. **WebSockets**: Utilize a WebSocket library to facilitate real-time communication among participants during a meeting.
b. **UI for Messaging**: Create a user interface to display messages and send notifications.

#### Backend
a. **WebSocket Server**: Implement a WebSocket server for handling real-time messaging and notifications.
b. **Message Storage**: Store messages in MongoDB to retrieve the chat history.

##### Code Snippet Example (WebSocket Server using Socket.io - Node.js/Express)
```js
const express = require('express');
const http = require('http');
const socketIo = require('socket.io');
const mongoose = require('mongoose');

const app = express();
const server = http.createServer(app);
const io = socketIo(server);

io.on('connection', (socket) => {
    console.log('a user connected:', socket.id);

    socket.on('send_message', (data) => {
        // Store message in MongoDB
        // Assume Message is your MongoDB model
        const newMessage = new Message(data);
        newMessage.save().catch((error) => console.log(error));

        // Broadcast message to other users in the same meeting
        socket.broadcast.to(data.meetingId).emit('receive_message', data);
    });

    socket.on('join_meeting', (meetingId) => {
        socket.join(meetingId);
    });

    socket.on('disconnect', () => {
        console.log('user disconnected:', socket.id);
    });
});

server.listen(3000, () => {
    console.log('listening on *:3000');
});
```

##### Code Snippet Example (WebSocket Client using Socket.io - React)
```jsx
import React, { useEffect, useState } from 'react';
import io from 'socket.io-client';

function Chat({ meetingId }) {
    const [messages, setMessages] = useState([]);
    const [inputMessage, setInputMessage] = useState('');
    const socket = io('http://localhost:3000');

    useEffect(() => {
        socket.emit('join_meeting', meetingId);

        socket.on('receive_message', (message) => {
            setMessages((prevMessages) => [...prevMessages, message]);
        });

        return () => {
            socket.disconnect();
        };
    }, [socket, meetingId]);

    const sendMessage = () => {
        const messageData = {
            text: inputMessage,
            timestamp: new Date(),
            userId: 'yourUserId',
            meetingId,
        };
        setMessages((prevMessages) => [...prevMessages, messageData]);
        socket.emit('send_message', messageData);
        setInputMessage('');
    };

    return (
        <div>
            <div>
                {messages.map((msg, idx) => (
                    <p key={idx}>{msg.text}</p>
                ))}
            </div>
            <input 
                type="text" 
                value={inputMessage} 
                onChange={(e) => setInputMessage(e.target.value)} 
            />
            <button onClick={sendMessage}>Send</button>
        </div>
    );
}

export default Chat;
```

#### Frontend & Backend
a. **Data Encryption**: Encrypt sensitive user data and messages using cryptography libraries.
b. **Authentication**: Implement secure JWT-based authentication and authorization.
c. **Secure Socket**: Use secure WebSocket (WSS) for data transmission during chats.

### 7. User Experience (UX) and User Interface (UI) Design
a. **Mobile Responsiveness**: Ensure your application is mobile-friendly and adaptable to various screen sizes.
b. **User Onboarding**: Develop user-friendly onboarding processes.
c. **Loading States**: Provide loading animations and feedback during any data fetching to keep the user informed.
d. **Error Handling**: Implement user-friendly error messages and options to troubleshoot or report issues.

### 8. Testing and Deployment
a. **Unit Testing**: Write tests for individual components and functions.
b. **Integration Testing**: Ensure different parts of the application work together smoothly.
c. **User Testing**: Gather feedback from real users about usability and improvements.
d. **Deployment**: Choose a cloud platform or a server for deploying your application, and set up continuous deployment.

Building a meeting scheduler application requires thoughtful consideration in ensuring user-friendliness, privacy, and real-time functionalities. These guidelines and code snippets should help you make a good start, but remember that application development is iterative, and consistently gathering feedback and making improvements is key to success! If you have more specific queries or need assistance on a particular topic, feel free to ask!

---

#### A. Interface Design
Creating an elegant, intuitive, and user-friendly interface is crucial for the success of any application. Consider the following aspects:

1. **Consistency**: Ensure that the UI design is consistent across all pages and components.
2. **Color Scheme**: Choose a color scheme that is visually appealing and aligns with any branding elements.
3. **Typography**: Ensure readability and aesthetic appeal by choosing a legible and stylish font.
4. **Icons and Images**: Use high-quality, descriptive icons and images.
   
#### B. Interactive Elements
1. **Calendar View**: 
   - Implement a calendar view where users can visually select days for scheduling meetings.
   - Highlight days with scheduled meetings and allow users to click on a day to view the details.

2. **Drag-and-Drop**: 
   - Allow users to reschedule meetings using a drag-and-drop interface within the calendar.
   
3. **Tooltips**: 
   - Use tooltips to provide additional information when hovering over specific elements.
   
#### C. Animations
Subtle animations can enhance user engagement:
1. **Transition Animations**: Add smooth transitions when switching between different views or loading data.
2. **Loading Animations**: Provide animated feedback while data is loading to keep the user informed.

#### D. Accessibility
Ensure that your UI is accessible to people with disabilities:
1. **Contrast**: Ensure that text is legible and there’s sufficient contrast between text and background colors.
2. **Keyboard Navigation**: Ensure all functionalities can be accessed via keyboard.
3. **Screen Reader Support**: Ensure your UI components are properly labeled for screen readers.

#### E. Responsiveness
Make sure your application UI adapts to various screen sizes and orientations:
1. **Grid Layout**: Use a flexible grid layout to allow content reorganization according to screen size.
2. **Media Queries**: Implement media queries to adjust styles for different screen sizes.
   
#### Code Snippet Example (React - Calendar Component)
```jsx
import React, { useState } from 'react';
import { Calendar, momentLocalizer } from 'react-big-calendar';
import moment from 'moment';
import 'react-big-calendar/lib/css/react-big-calendar.css';

const localizer = momentLocalizer(moment);

function SchedulerCalendar({ events }) {
    const [currentDate, setCurrentDate] = useState(new Date());

    const handleSelect = ({ start, end }) => {
        // Handle date selection logic here
    };

    const eventStyleGetter = (event) => {
        // Customize event styles here
        return {
            style: {
                backgroundColor: event.isImportant ? 'red' : 'blue',
            }
        };
    };

    return (
        <Calendar
            localizer={localizer}
            defaultDate={currentDate}
            defaultView="month"
            events={events}
            style={{ height: "80vh" }}
            onSelectSlot={handleSelect}
            selectable={true}
            eventPropGetter={eventStyleGetter}
        />
    );
}

export default SchedulerCalendar;
```

In the above code, you should have an `events` prop, which is an array of event objects passed to the `SchedulerCalendar` component. Each event object should at least contain `start`, `end`, and `title` properties. The `react-big-calendar` library provides a highly customizable calendar view, and `moment` is used for date management.

Creating a rich UI involves multiple aspects, such as providing a visually appealing design, ensuring smooth interactions, and enabling accessibility. A well-designed UI not only enhances user satisfaction but also positively impacts user adoption and retention. If you have specific aspects or technologies you would like to explore further, please let me know!

---

### 10. AWS Deployment of the Meeting Scheduler App

Deploying your meeting scheduler application to AWS involves several key steps. AWS offers a suite of tools and services that can facilitate efficient and scalable deployment.

#### A. Set Up AWS Account
If not already done, create an AWS account, set up AWS IAM (Identity and Access Management) users, and configure the necessary permissions.

#### B. Prepare the App for Deployment
Ensure that your application is ready for deployment:
1. **Testing**: Confirm that all functionality is working as expected.
2. **Optimization**: Ensure your code is optimized for performance.
3. **Environment Variables**: Set up environment variables for API keys, database URIs, etc., for security.

#### C. Database Migration to AWS
Migrate your local database to an AWS service like Amazon RDS (Relational Database Service):
1. **Create an RDS Instance**: Choose the desired DB engine (e.g., PostgreSQL, MySQL) and configure the instance according to your requirements.
2. **Migrate Data**: Utilize tools like AWS Database Migration Service (DMS) to migrate your local database to RDS.
3. **Update Connection Strings**: Adjust your app's database connection strings to point to the new RDS instance.

#### D. Backend API Deployment
Deploy your backend API using AWS Elastic Beanstalk or AWS ECS (Elastic Container Service) if utilizing containers:
1. **Prepare API for Deployment**: Package your API and ensure that all dependencies are listed.
2. **Deploy API**: Use AWS Elastic Beanstalk for a straightforward PaaS deployment or ECS for containerized applications.
3. **Configure**: Set up environment variables, VPC (Virtual Private Cloud), and security groups for secure access.

#### E. Frontend App Deployment
Deploy your frontend application using AWS Amplify or AWS S3 + AWS CloudFront for static site hosting:
1. **Build**: Generate the production build of your frontend application.
2. **Deploy using AWS Amplify**: Utilize Amplify Console to connect to your repository and deploy the application, taking advantage of features like CI/CD, custom domain setup, and HTTPS.
   OR
   **Deploy using AWS S3 + CloudFront**: Host static files on S3 and use CloudFront to deliver the content through a CDN, ensuring reduced latency and increased loading speeds globally.

#### F. DNS Configuration
Configure DNS to point to your deployed applications:
1. **Acquire a Domain**: If you don’t have one, purchase a domain using Amazon Route 53 or another domain registrar.
2. **Configure DNS Records**: Point your domain/subdomain to the AWS resources (e.g., Elastic Beanstalk environment URL, CloudFront distribution) using Route 53.

#### G. Monitor & Optimize
Leverage AWS CloudWatch for monitoring and ensure optimal performance:
1. **Monitoring**: Set up CloudWatch Alarms for resource utilization monitoring and anomaly detection.
2. **Optimization**: Analyze metrics and optimize resource allocations based on usage patterns.

#### Example AWS CLI Commands
- Deploy a Dockerized app to ECS:
  ```sh
  aws ecs create-cluster --cluster-name MyCluster
  aws ecs register-task-definition --cli-input-json file://task-definition.json
  aws ecs create-service --cluster MyCluster --service-name MyService --task-definition MyTaskDefinition --desired-count 1 --launch-type FARGATE --network-configuration "awsvpcConfiguration={subnets=[subnet-xxxxxx],securityGroups=[sg-xxxxxx],assignPublicIp=ENABLED}"
  ```
- Copy frontend build to S3:
  ```sh
  aws s3 sync build/ s3://my-bucket-name/
  ```
  Ensure to replace placeholder values with actual values pertaining to your AWS setup and application.

#### Final Thoughts
Deployment involves several steps, and each one is crucial to ensure that the application runs smoothly in the production environment. Always validate the functionality post-deployment and closely monitor the application to address any issues promptly. Further enhancements, like implementing Auto Scaling, can be considered for managing varying loads efficiently.

**Comment by tinapatil13** [comment]
[Component.zip](https://github.com/akash-coded/mern/files/12829120/Component.zip)
Please find zip folder, i am facing issue in backend for login sign up page, help me to resolve by giving solution with explanations.
